### ISSUE TITLE
KeyStore usage in Corda

### ISSUE THREAD

**Issue by remmeier** [corda/corda#5832]
We upgraded to Corda 4.3 resp. 4.4 master together with JDK 11. Mostly things seem great. Thx for all the work. Especially being able to move to JDK 11 is really helpful. Unfortunate exception is the transaction throughput which dropped by 50% for some of our flows at least for the OSS-flavor of Corda (we rather hoped to gain 20% due to the new JDK). 

One reason is documented in https://github.com/corda/corda/issues/5831. A second reason seems to be related to the X509KeyStore implementation:

![image](https://user-images.githubusercontent.com/8008853/71558892-3ac99700-2a58-11ea-9ad3-0f6e8a544f8e.png)

The implementation did not change. So it probably is related to the underlying JDK implementation. In this case we made use of Azul Zulu 11 to be sure to get that Corda JDK 11 fix.  That 30% performance  hit that can be seen just to read the private key from the key store is a bit a lot :-(  We have not seen this behavior in previous versions with AdoptOpenJdk 8. So maybe that is something Corda would like to guard itself by caching the instantiated key material.

Related to this we also noticed that instantiating PublicKeys is rather expensive in general (but not THAT expensive as in the trace above). So potentially `PublicKeySerializer` should cache the deserialized key material.  

After fixing those things locally we started to see 20% performance improvements in some first areas. Can do a PR if desired.

**Comment by rick-r3** [comment]
Do note that whilst Corda can be compiled for Java 11, we have not yet finished the certification work, so we do not recommend or support running on Java 11.  We are also unaware of any released versions of JDK11 that are fully capable of running Corda.  We have outstanding bug fixes that we are waiting on from JDK11.

Thanks for the analysis on the `KeyStore`.  We'll take a look at what is happening there in our own performance engineering nodes and look to make improvements if we can reproduce the above.

A caching `PublicKeySerializer` has always been in the pipeline.  We just didn't quite get to it yet.

**Comment by remmeier** [comment]
we started trying it out based on the advertisement at CordaCon. but so for production systems we sticked to JDK 8. But thank you for the warning.

As for `PublicKeySerializar` we have this here: https://github.com/remmeier/corda/commit/84c445f3ed6fbd9aa3e395d617c73f11137a3f35 It caches the public key with SHA256. That seems much faster than creating the public key. Did not create a PR yet and not quite sure about it. It is a bit improvised due to the limited classpath of the deterministic JVM (meaning no access to apache commons LRU map). The solution might either be ok or can be simplified...
